tests/Metadata/Property/Factory/ConstructorPropertyMetadataFactoryTest.php

```diff
@@ -0,0 +1,160 @@
+            'collection_operation_name' => 'hello',
+        ]);
+        $this->assertFalse($fooMetadata->isWritable());
+    }
+
+    public function testItCreateAndSetRightMetadataIfNoFactoryGiven()
+    {
+        $factory = new ConstructorPropertyMetadataFactory($this->getMethodResolverMock());
+        $fooMetadata = $factory->create(DummyObjectWithConstructor::class, 'bar', [
+            'collection_operation_name' => 'hello',
+        ]);
+
+        $this->assertTrue($fooMetadata->isWritable());
+    }
+
+    public function testItDoesntThrowErrorForAnyKindOfObject()
+    {
+        $propertyMetadataFactory = $this->prophesize(PropertyMetadataFactoryInterface::class);
+
+        $type = new Type(Type::BUILTIN_TYPE_STRING);
+        $fooMetadata = new PropertyMetadata($type, 'foo', true, true, false, false, true, false, 'http://example.com/foo', null, ['foo' => 'bar']);
+        $propertyMetadataFactory->create(DummyObjectWithoutConstructor::class, 'foo', Argument::any())->willReturn($fooMetadata)->shouldBeCalled();
+
+        $factory = new ConstructorPropertyMetadataFactory(
+            $this->getMethodResolverMock(),
+            $propertyMetadataFactory->reveal()
+        );
+
+        $fooMetadata = $factory->create(DummyObjectWithoutConstructor::class, 'foo', [
+            'collection_operation_name' => 'hello',
+        ]);
+        $this->assertTrue($fooMetadata->isWritable());
+    }
+
+    public function testItReturnTheOriginalPropertyIfNoOperationFound()
+    {
+        $propertyMetadataFactory = $this->prophesize(PropertyMetadataFactoryInterface::class);
+
+        $type = new Type(Type::BUILTIN_TYPE_STRING);
+        $fooMetadata = new PropertyMetadata($type, 'foo', true, false, false, false, true, false, 'http://example.com/foo', null, ['foo' => 'bar']);
+        $propertyMetadataFactory->create(DummyObjectWithoutConstructor::class, 'foo', Argument::any())->willReturn($fooMetadata)->shouldBeCalled();
+
+        $resolver = $this->prophesize(OperationMethodResolverInterface::class);
+        $resolver->getCollectionOperationMethod(Argument::cetera())->willThrow(new RuntimeException());
+
+        $factory = new ConstructorPropertyMetadataFactory(
+            $resolver->reveal(),
+            $propertyMetadataFactory->reveal()
+        );
+
+        $newMetadata = $factory->create(DummyObjectWithoutConstructor::class, 'foo', [
+            'collection_operation_name' => 'hello',
+        ]);
+        $this->assertFalse($fooMetadata->isWritable());
+        $this->assertEquals(spl_object_hash($fooMetadata), spl_object_hash($newMetadata));
+    }
+
+    private function getMethodResolverMock($method = 'post')
+    {
+        $resolver = $this->prophesize(OperationMethodResolverInterface::class);
+        $resolver->getCollectionOperationMethod(Argument::any(), Argument::any())->willReturn($method);
+
+        return $resolver->reveal();
+    }
+}
+
+class DummyObjectWithConstructor
```

Comment:
Wasn't best practice to put these in another file?

Reply:
Well, that depends on your POV. IMO re-using dummy classes in many test is not a good practice because it means your test is dependent on something _outside_ (see [FIRST](https://stackoverflow.com/questions/18024785/tdd-first-principle)). I do it only in case I need doctrine entities to avoid configuration duplication.

TL;DR: I don't think it's a big deal.

Reply:
I agree but it's how we choose to do in this project and we have to follow the project rules :).

Reply:
I know for a fact that I did the same as you and I was told to move my classes out of the test file that's all :p.

Reply:
So I guess it's a rule :) .

Reply:
It's a rule, because it's what recommended is by the PSRs. Anyway, we try to enforce **not** reusing dummy classes, because of first and - especially - because it's a pain to maintain such tests.

Both can be done together: a test = a dummy class in a separate file. It's what we should enforce.

Reply:
Couldn't an anonymous class be used ?

Reply:
Putting them outside doesn't mean reusing them. We should not have more than 1 class in a file, as that wouldn't work for PSR-4 autoloading.